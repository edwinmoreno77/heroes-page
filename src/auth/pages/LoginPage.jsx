import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";


export const LoginPage = () => {

    const { login } = useContext(AuthContext);

    const navigate2 = useNavigate();

    const onLogin = () => {

        const lastPath = localStorage.getItem("lastPath") || "/";

        login('Edwin Moreno');

        navigate2(lastPath, {
            replace: true
        });
    }
    return (
        <div className="m-5">
            <h1>Login Page</h1>
            <hr />

            <button
                onClick={onLogin}
                className="btn btn-primary">
                Login
            </button>
            <div className="row">
                <div className="col-4">
                    <img src="/assets/img/JusticeLeague-Backgrounds2.jpg" alt="" />
                </div>
            </div>


        </div>
    )
}
