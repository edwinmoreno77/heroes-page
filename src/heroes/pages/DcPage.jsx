import { HeroList } from "../components";


export const DcPage = () => {
    return (
        <>

            <h1 className="mt-5">Dc Page</h1>
            <hr />
            <HeroList publisher={'DC Comics'} />


        </>
    )
}
