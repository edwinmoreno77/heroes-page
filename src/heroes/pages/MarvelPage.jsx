import { SliderMarvel } from "../../ui/components";
import { HeroList } from "../components";




export const MarvelPage = () => {
    return (
        <>
            <SliderMarvel />
            <div className="m-3">
                <HeroList publisher={'Marvel Comics'} />
            </div>
        </>
    )
}
