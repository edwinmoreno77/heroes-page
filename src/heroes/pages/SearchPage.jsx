import { useLocation, useNavigate } from "react-router-dom";

import queryString from "query-string";

import { useForm } from "../../hooks/useForm"
import { HeroCard } from "../components/HeroCard"
import { getHeroesByName } from "../helpers";



export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const heroes = getHeroesByName(q);

    const showSearch = (q.length === 0);
    const showError = (q.length > 0) && heroes.length === 0;


    const { searchText, onInputChange } = useForm({
        searchText: q
    });

    const onSearchSubmit = (e) => {
        e.preventDefault();
        // if (searchText.trim().length <= 1) return;
        navigate(`?q=${searchText}`);
    }



    return (
        <>
            <div className="row searchImgBg">

                <div className="col-8 m-5">
                    <h4>Searching</h4>
                    <form onSubmit={onSearchSubmit}>
                        <input
                            className="form-control shadow mt-5"
                            placeholder="Search a hero"
                            type="text"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={onInputChange}
                        />

                        <button className="btn btn-outline-primary mt-2 shadow">
                            Search
                        </button>
                    </form>

                    {/* <div className="m-1">
                        <img src="/assets/img/JusticeLeague-Backgrounds2.jpg" alt="" />
                    </div> */}


                </div>

                <div className="container">
                    <div className="row row-cols-1 row-cols-md-3 row-cols-lg-4 row-cols-xl-5 g-2 m-5 p-5">
                        {/* {
                        (q === '')
                            ? <div className="alert alert-warning">Search a hero</div>
                            : (heroes.length === 0)
                            && <div className="alert alert-danger">No hero with <b>{q}</b></div>
                    } */}
                        <div className=" mt-5 alert alert-primary animate__animated animate__fadeIn"
                            style={{ display: showSearch ? '' : 'none' }}>
                            Search a hero
                        </div>

                        <div className="mb-5 alert alert-danger animate__animated animate__fadeIn"
                            style={{ display: showError ? '' : 'none' }}>
                            No hero with <b>{q}</b>
                        </div>

                        {
                            heroes.map(hero => (
                                <HeroCard key={hero.id} {...hero} />
                            ))
                        }
                    </div>

                </div>

            </div>

        </>
    )
}
