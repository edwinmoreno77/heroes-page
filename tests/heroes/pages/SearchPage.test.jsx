import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, useNavigate } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";


const mockdUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockdUseNavigate,
}));


describe('<SearchPahe/> test', () => {

    beforeEach(() => jest.clearAllMocks());

    test('should show component', () => {

        const { container } = render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        )


        expect(container).toMatchSnapshot();

    });


    test('should show Batman and input with value queyString', () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        // screen.debug();

        const input = screen.getByRole('textbox');
        expect(input.value).toBe('batman');

        const img = screen.getByRole('img');
        expect(img.src).toContain('dc-batman.jpg');

        const div = screen.getByTestId('div-display-none');
        expect(div.style).toHaveProperty('display', 'none');
        //other way
        expect(div.style.display).toBe('none');

    });

    test("it should show an error if it doesn't find the hero", () => {

        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        )

        const div = screen.getByTestId('div-display-none');
        expect(div.style.display).toBe('');

    });

    test('you must call navigate to the new screen', () => {

        const inputValue = 'batman123';

        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        )

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { value: inputValue } });

        const form = container.querySelector('form');
        fireEvent.submit(form);

        expect(mockdUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);

    });
});