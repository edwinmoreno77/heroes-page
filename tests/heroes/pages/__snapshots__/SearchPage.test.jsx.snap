// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPahe/> test should show component 1`] = `
<div>
  <div
    class="row searchImgBg"
  >
    <div
      class="col-8 m-5"
    >
      <h4>
        Searching
      </h4>
      <form>
        <input
          autocomplete="off"
          class="form-control shadow mt-5"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-2 shadow"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="container"
    >
      <div
        class="row row-cols-1 row-cols-md-3 row-cols-lg-4 row-cols-xl-5 g-2 m-5 p-5"
      >
        <div
          class=" mt-5 alert alert-primary animate__animated animate__fadeIn"
        >
          Search a hero
        </div>
        <div
          class="mb-5 alert alert-danger animate__animated animate__fadeIn"
          data-testid="div-display-none"
          style="display: none;"
        >
          No hero with 
          <b />
        </div>
      </div>
    </div>
  </div>
</div>
`;
